### ISSUE TITLE
Check for existing users first by `employeeId` (instead of `userPrincipalName`)

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/women-for-women#7]
## Background, context, and business value
Currently, the mapping employed between BambooHR -> Azure relies on the `Work Email` entered into BambooHR to assign `userPrincipalName` in Azure. 

This is not ideal though, as the W4W administrator entering data into BambooHR's `Work Email` field might make a mistake with his/her data entry and doing so could cause severe complications in the Azure environment.

To correct this, we must explore alternative ways to generate/check for `userPrincipalName` that reduces the possibility for entry error. The client has also shared that `employeeId` is the better unique identifier to use between the Bamboo and Azure systems.

## The specific request, in as few words as possible
Use Azure's `employeeId` field to check for existing users. Then compare employees' `userPrincipalName` to determine if users are unique. 
1. Modify the Step 1 in the automation, to list _all_ users (not only those with matching `userPrincipalName`). API docs: https://docs.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0&tabs=http
2. From this list of users, find the user where `employeeId` matches the BambooHR `Employee #` from state

Compare the BambooHR employee data with the Azure user record...
1. If `employeeId` AND `userPrincipalName` are matching --> update user in Azure
2. If `employeeId` and `userPrincipalName` do NOT match --> stop automation flow and throw error to force run failure & notify admins that `Employee Id and User Principal Name do not match. Please review this user to confirm the Work Email entered in BambooHR.`
3.  If `employeeId` not found in list of existing users --> create new user in Azure
 
## state.json
See LP for `Microsoft (Women for Women)` Azure test instance login. 
Use `stateMembershipUpdates.json`. 

## adaptor
`language-http`

## expression.js
Update `Step 1. Get user` to list ALL users and then search for users with matching `employeeId`
https://github.com/OpenFn/women-for-women/blob/36822736e49f7b01d978c1e15b7a4de6dbb39870/jobs/upsertUser.js#L330

## output.json
Testing with the following states should produce these results...
1. `stateMembershipUpdates.json` has employee data where both `employeeId` does not exist --> should create a new user
2. `stateMembershipUpdates.json` run this state again --> should update the user created in the prior run
2. `stateEmployeeNewEmail.json` has employee data where `employeeId` matches with an existing Azure user, but the `work email`/ `userPrincipalName` does NOT match --> should stop the run and log message, no actions made in Azure

**Comment by lakhassane** [comment]
@aleksa-krolls @jedbgold on this. Logic has been implemented. But I am unable to test. I am having an error as: Authorization_RequestDenied (as well for the version in prod right now). I can't seem to have access on Azure either. Could it be related?

**Comment by jedbgold** [comment]
@lakhassane thanks for the heads up. Now that you have access to Azure again, let us know how the testing goes when you can.

**Comment by aleksa-krolls** [comment]
@lakhassane So far looking good, but would it be possible to force a run _failure_ and throw an error when `Employee Id and User Principal Name do not match. Please review this user to confirm the Work Email entered in BambooHR`? 

In such scenarios we want to draw the admin's attention and [this successful run](https://www.openfn.org/projects/2976/runs/10406777) raises no alarm... and if this scenario occurs, the admin will need to review the users in Azure and maybe change the email address provided by BambooHR. We could maybe trigger an email alert instead but wondering if it would easier to force a run failure given that this is a human data entry error preventing a successful data sync. What do you and @taylordowns2000 recommend?

**Comment by lakhassane** [comment]
@aleksa-krolls as suggested we throw an error when `upn` do not match  `employeeId` and at the same time we log the message for admin. So it's all good.

**Comment by aleksa-krolls** [comment]
@lakhassane Great, thank you! 
@jedbgold [Here](https://www.openfn.org/projects/2976/runs/10409351) is an example where this error is thrown if the `userPrincipalName` does not match. Please test the automation end-to-end to ensure this _and_ existing actions are working. I'd add a test scenario to our test suite to test for this scenario where Bamboo `Employee #` matches Azure `employeeId`, but the `Work Email` and `userPrincipalName` do not match.

**Comment by aleksa-krolls** [comment]
@lakhassane This is priority, please pick this up next. 
## Issue
Sometime `employeeId` has not been added in Azure. Therefore we need to check `userPrincipalName` only when we upsert to determine if we should create or update a record. Currently if there is no  `employeeId` , we wrongly assume it's a new user and this is causing runs to fail: https://openfn.org/projects/p56pxp/runs/rrkax63g

## Change request
Please amend the logic originally specified above request (see bold for changes we need to implement). 
Compare the BambooHR employee data with the Azure user record...
**0. If no Azure `employeeId` is specified, check if `userPrincipalName` match --> if yes, update user in Azure. 
If no --> create new user.** 
1. If `employeeId` AND `userPrincipalName` are matching --> update user in Azure
2. If `employeeId` and `userPrincipalName` do NOT match --> stop automation flow and throw error to force run failure & notify admins that Employee Id and User Principal Name do not match. Please review this user to confirm the Work Email entered in BambooHR.
3. If `employeeId` not found in list of existing users --> create new user in Azure

## Logic to update
https://github.com/OpenFn/women-for-women/blob/d3a7a129f81cd7b66563c23ae2124645f61a4a39/jobs/upsertUser.js#L609-L627

## To test
Re-try this run should succeed & update the Azure user: https://openfn.org/projects/p56pxp/runs/rrkax63g
This an example employee where there is _no_ Azure `employeeId`, but the `userPrincipalName` matches... so this user should be updated.

**Comment by aleksa-krolls** [comment]
@lakhassane One more small change request to add criteria to compare work email before updating a user...
## Request
Add logic to verify that both the `userPrincipalName` match (existing logic) **_AND_** that...
`email` match before updating an Azure user. 

## expression
I think it's just a quick update to the logic on L618 to also compare Azure `mail` with Bamboo `Work Email`
https://github.com/OpenFn/women-for-women/blob/d3a7a129f81cd7b66563c23ae2124645f61a4a39/jobs/upsertUser.js#L618-L619

**Comment by aleksa-krolls** [comment]
@lakhassane So I'm testing the changes you made for paging to ensure we can always upsert users properly. It works great here (https://openfn.org/projects/p56pxp/runs/rj5k84zz) and finds the existing Matias user. However, in this run (https://openfn.org/projects/p56pxp/runs/ryajy6e8) I'm running into that error again where it doesn't find an existing user to update, but rather creates: `"Another object with the same value for property userPrincipalName already exists"`.

Why does the integration not recognize the existing `Russel` Azure user and tries to create rather than update? 
https://portal.azure.com/?feature.argsubscriptions=true&feature.iris=true&feature.irisalt=false&feature.prefetchtokens=false&feature.showservicehealthalerts=true&feature.usemsallogin=true&sessionId=36224c8c55d84f7098aacd78636e0de3#blade/Microsoft_AAD_IAM/UserDetailsMenuBlade/Profile/userId/665b2dd0-bb7e-462a-b003-5814f0e300a9
![image.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/e31ff5e6-a142-451e-9736-69164808915e)

To recap, we want to make sure the following logic will work for upserts...
1. If Bamboo employee records matches with **both** Azure UPN and email, update Azure record. (Adding the email criteria is what we just implemented.)
2. If Bamboo employee record does _not_ match **neither** Azure UPN nor mail in Azure, create a new user record.
3, If Bamboo employee record only matches Azure UPN or email, throw message and log error details. 

## To test
Use this for state: https://openfn.org/projects/p56pxp/messages/mqpqwgkz

**Comment by aleksa-krolls** [comment]
@lakhassane To pick up tomorrow, I see a `TypeError` here: https://openfn.org/projects/p56pxp/runs/r5eazz7a

**Comment by aleksa-krolls** [comment]
I'm still seeing a `TypeError` on L882 of this run log: https://openfn.org/projects/p56pxp/runs/rpdabby7

**Comment by aleksa-krolls** [comment]
@lakhassane This change is looking good. But I'm finding some other `BadRequest` errors that seems to be finding duplicates in the test environment, but that does not seem right...

1. `BadRequest` conflict when trying to add a `Member` for a user... don't we always upsert `member` records? Why is a dupe being flagged? 
https://www.openfn.org/projects/p56pxp/runs/rj5k3med
2. Other `member` issue examples: 
https://www.openfn.org/projects/p56pxp/runs/raa3nd9n
https://www.openfn.org/projects/p56pxp/runs/rmkyxeny
https://www.openfn.org/projects/p56pxp/runs/rj5k3mbj
3. Seems to find duplicate Managers? But I only see 1 for `ladams` when I check on the front-end: https://www.openfn.org/projects/p56pxp/runs/rk57bw3a
![image](https://user-images.githubusercontent.com/16758106/105060032-98c08680-5a80-11eb-82cc-14d466834da5.png)

**Comment by lakhassane** [comment]
@taylordowns2000 can we chat about `4` tomorrow morning?

**Comment by aleksa-krolls** [comment]
@lakhassane Would it be easy enough to push any changes you make also to this job `updateUserProd.js`? This is the exact same job BUT I removed the `#EXT...` suffix we add to `userPrincipalName` when connecting with the test environment. Unfortunately because of this difference, I think we'll need to keep 1 production and 1 sandbox copy of the job. 
https://github.com/OpenFn/women-for-women/blob/master/jobs/upsertUserProd.js

**Comment by lakhassane** [comment]
Not that hard @aleksa-krolls. Mostly those are small changes so I can duplicate any future change on this job as well.

**Comment by aleksa-krolls** [comment]
@lakhassane Keep me posted on the approach you ultimately decide to implement for my feedback `1`-`4`? Keep me posted if easier to discuss over a call, and happy to jump on a quick call this afternoon. 
cc @taylordowns2000

**Comment by aleksa-krolls** [comment]
@taylordowns2000 @lakhassane This run is still timing out. What's the recommendation here? Can we set this up as a long-running job? (The triggering message is for a pretty large record set, so I don't anticipate the client needing this long-running job more than a couple times per year... but would be great to ensure we can handle occasional bulk record updates?)
https://www.openfn.org/projects/p56pxp/messages/mqpqwgkz

**Comment by taylordowns2000** [comment]
Heyo, this is already set to long run (and the last run failed after 235 seconds—indicating that it was properly given the extra time needed to run as a "long run".) Check this out: https://www.openfn.org/projects/p56pxp/runs/r89nyxde

**Comment by lakhassane** [comment]
@aleksa-krolls could you point to the run that timed out? The one from 24.01 did not timed out but failed due to normal conditions (`work email not provided`, ...)

**Comment by aleksa-krolls** [comment]
@lakhassane sorry i see the latest run is not timing out... my bad, was looking at older runs. Thanks!
